File: .web/pages/globalping.js
/** @jsxImportSource @emotion/react */


import { Fragment, useCallback, useContext } from "react"
import { EventLoopContext, StateContexts } from "/utils/context"
import { Event, getBackendURL, isTrue } from "/utils/state"
import { Button as RadixThemesButton, Container as RadixThemesContainer, Dialog as RadixThemesDialog, DropdownMenu as RadixThemesDropdownMenu, Flex as RadixThemesFlex, Heading as RadixThemesHeading, Link as RadixThemesLink, Text as RadixThemesText } from "@radix-ui/themes"
import env from "/env.json"
import NextLink from "next/link"
import { SkeletonText, Stack } from "@chakra-ui/react"
import { Grid as DataTableGrid } from "gridjs-react"
import "gridjs/dist/theme/mermaid.css"
import NextHead from "next/head"



export function Fragment_e1d8101aef680242d1bb09a07fe00379 () {
  const state__form_input_state = useContext(StateContexts.state__form_input_state)


  return (
    <Fragment>
  {isTrue(state__form_input_state.loading) ? (
  <Fragment>
  <Stack sx={{"width": "100%"}}>
  <SkeletonText noOfLines={25}/>
</Stack>
</Fragment>
) : (
  <Fragment>
  <RadixThemesContainer css={{"height": "0px"}}/>
</Fragment>
)}
</Fragment>
  )
}

export function Fragment_1762bb90abdb81b879b2a22edbbe01a1 () {
  const [addEvents, connectError] = useContext(EventLoopContext);


  return (
    <Fragment>
  {isTrue(connectError !== null) ? (
  <Fragment>
  <RadixThemesDialog.Root open={connectError !== null}>
  <RadixThemesDialog.Content>
  <RadixThemesDialog.Title>
  {`Connection Error`}
</RadixThemesDialog.Title>
  <RadixThemesText as={`p`}>
  {`Cannot connect to server: `}
  {(connectError !== null) ? connectError.message : ''}
  {`. Check if server is reachable at `}
  {getBackendURL(env.EVENT).href}
</RadixThemesText>
</RadixThemesDialog.Content>
</RadixThemesDialog.Root>
</Fragment>
) : (
  <Fragment/>
)}
</Fragment>
  )
}

export function Fragment_7dcc98a07bb06ecb7f2659a550743607 () {
  const state__form_input_state = useContext(StateContexts.state__form_input_state)


  return (
    <Fragment>
  {isTrue(state__form_input_state.show_data) ? (
  <Fragment>
  <DataTableGrid columns={state__form_input_state.ping_results.columns} data={state__form_input_state.ping_results.data} search={true} sort={true}/>
</Fragment>
) : (
  <Fragment>
  {isTrue(state__form_input_state.show_initial) ? (
  <Fragment>
  <RadixThemesContainer css={{"backgroundColor": "#f4f5f8", "borderRadius": "7px", "marginTop": "20px"}}>
  <RadixThemesText as={`p`} css={{"padding": "1em", "textAlign": "center"}}>
  {`Your ping results will be displayed here!`}
</RadixThemesText>
</RadixThemesContainer>
</Fragment>
) : (
  <Fragment>
  <RadixThemesContainer/>
</Fragment>
)}
</Fragment>
)}
</Fragment>
  )
}

export function Button_e1860d1c801ee3ad17ecffb1c05d2aa4 () {
  const [addEvents, connectError] = useContext(EventLoopContext);

  const on_click_347c0269d48fabdaf2e9480e861a7972 = useCallback((_e) => addEvents([Event("state.form_input_state.handle_submit", {})], (_e), {}), [addEvents, Event])

  return (
    <RadixThemesButton onClick={on_click_347c0269d48fabdaf2e9480e861a7972} size={`3`}>
  {`Run Globalping Test!`}
</RadixThemesButton>
  )
}

export default function Component() {

  return (
    <Fragment>
  <Fragment_1762bb90abdb81b879b2a22edbbe01a1/>
  <Fragment>
  <RadixThemesFlex align={`center`} css={{"position": "fixed", "top": "0px", "backgroundColor": "white", "padding": "1em", "height": "4em", "width": "100%", "zIndex": "5", "borderBottom": "0.1px solid #eae7ec", "flexDirection": "row"}} gap={`2`}>
  <RadixThemesFlex align={`start`} css={{"flexDirection": "row"}} gap={`2`}>
  <img css={{"width": "2em"}} src={`/favicon.ico`}/>
  <RadixThemesHeading css={{"fontSize": "2em"}}>
  {`App`}
</RadixThemesHeading>
</RadixThemesFlex>
  <RadixThemesFlex css={{"flex": 1, "justifySelf": "stretch", "alignSelf": "stretch"}}/>
  <RadixThemesFlex align={`center`} css={{"flexDirection": "row"}} gap={`5`}>
  <RadixThemesDropdownMenu.Root>
  <RadixThemesDropdownMenu.Trigger>
  <RadixThemesButton css={{"backgroundColor": "black", "color": "white"}} radius={`medium`} variant={`soft`}>
  {`Menu`}
</RadixThemesButton>
</RadixThemesDropdownMenu.Trigger>
  <RadixThemesDropdownMenu.Content>
  <RadixThemesDropdownMenu.Item>
  {`item 1`}
</RadixThemesDropdownMenu.Item>
  <RadixThemesDropdownMenu.Item>
  {`Item 2`}
</RadixThemesDropdownMenu.Item>
  <RadixThemesDropdownMenu.Item>
  {`Item 3`}
</RadixThemesDropdownMenu.Item>
</RadixThemesDropdownMenu.Content>
</RadixThemesDropdownMenu.Root>
</RadixThemesFlex>
</RadixThemesFlex>
  <Fragment>
  <RadixThemesContainer css={{"backgroundColor": "#f4f5f8", "padding": "50px 1em 100px 1em"}}>
  <RadixThemesHeading align={`left`} css={{"marginTop": "10vh"}} size={`9`}>
  {`Get your ping from anywhere! 🌎`}
</RadixThemesHeading>
</RadixThemesContainer>
  <RadixThemesContainer css={{"backgroundColor": "#f4f5f8", "paddingLeft": "1em", "paddingRight": "1em"}}>
  <RadixThemesFlex align={`start`} css={{"backgroundColor": "#f4f5f8", "flexDirection": "row"}} gap={`2`}>
  <RadixThemesLink asChild={true} css={{"color": "rgba(34,46,58,.5)", "backgroundColor": "hsla(0,0%,100%,.5)", "padding": "15px", "borderRadius": "7px 7px 0px 0px"}}>
  <NextLink href={`/`} passHref={true}>
  {`My IP Address`}
</NextLink>
</RadixThemesLink>
  <RadixThemesLink asChild={true} css={{"color": "black", "backgroundColor": "white", "padding": "15px", "borderRadius": "7px 7px 0px 0px"}}>
  <NextLink href={`/globalping`} passHref={true}>
  {`My Global Ping`}
</NextLink>
</RadixThemesLink>
  <RadixThemesLink asChild={true} css={{"color": "rgba(34,46,58,.5)", "backgroundColor": "hsla(0,0%,100%,.5)", "padding": "15px", "borderRadius": "7px 7px 0px 0px"}}>
  <NextLink href={`https://reflex.dev`} passHref={true}>
  {`Track My Packet`}
</NextLink>
</RadixThemesLink>
</RadixThemesFlex>
</RadixThemesContainer>
  <RadixThemesFlex align={`center`} css={{"margin": "1em", "display": "flex", "alignItems": "center", "justifyContent": "center"}}>
  <RadixThemesContainer css={{"marginTop": "3em", "width": "100%"}}>
  <RadixThemesContainer css={{"height": "20px"}}/>
  <RadixThemesHeading>
  {`How does this work?`}
</RadixThemesHeading>
  <RadixThemesContainer css={{"height": "20px"}}/>
  <RadixThemesText as={`p`} css={{"wrap": "wrap", "width": "100%"}}>
  {`
                    Using the awesome features of jsDelivr Globalping, this cool web tool will randomly pick 100 spots across 
                    the globe and gather the ping times (in milliseconds) from your current IP address. To get started, just click "Run Globalping Test." 
                    It only takes around 10 seconds to gather and show your results. Feel free to run it again whenever you want!
                    `}
</RadixThemesText>
  <RadixThemesContainer css={{"height": "20px"}}/>
  <Button_e1860d1c801ee3ad17ecffb1c05d2aa4/>
  <RadixThemesContainer css={{"height": "20px"}}/>
  <Fragment_e1d8101aef680242d1bb09a07fe00379/>
  <Fragment_7dcc98a07bb06ecb7f2659a550743607/>
</RadixThemesContainer>
</RadixThemesFlex>
</Fragment>
</Fragment>
  <NextHead>
  <title>
  {`Reflex App`}
</title>
  <meta content={`A Reflex app.`} name={`description`}/>
  <meta content={`favicon.ico`} property={`og:image`}/>
</NextHead>
</Fragment>
  )
}

File: IP_website/globalping.py
# Name: Eric Dennis
# Started: 3/4/2024
# Last Updated: 3/4/2024
# Update Notes: 

### IMPORTS ###

import time
import json
import asyncio
import requests
import pycountry
import pandas as pd
import reflex as rx
import myipaddress as myip


def globalping():
        
    url = "https://api.globalping.io/v1/measurements"
    
    data = {
        "type": "ping",
        "target": f"{myip.public_ip()}",
        "locations": [
            {
            "magic": "world",
            "limit": 100
            }
        ]
    }
    
    headers = {"Content-Type": "application/json"}

    response = requests.post(url, json=data, headers=headers)
    measurement_id = response.json().get('id')
    response.raise_for_status()
    time.sleep(10)
    result_response = requests.get(f"https://api.globalping.io/v1/measurements/{measurement_id}")
    result_data = json.loads(result_response.text)

    # Extracting required data into lists.
    continents = []
    regions = []
    countries = []
    countries_full = []
    cities = []
    min_values = []
    max_values = []
    avg_values = []
    total = []
    loss = []
    rcv = []
    drop = []

    # Inserting results into their appropriate lists.
    for result in result_data['results']:
        continents.append(result['probe']['continent'])
        regions.append(result['probe']['region'])
        countries.append(result['probe']['country'])
        cities.append(result['probe']['city'])
        min_values.append(result['result']['stats']['min'])
        max_values.append(result['result']['stats']['max'])
        avg_values.append(result['result']['stats']['avg'])
        total.append(result['result']['stats']['total'])
        rcv.append(result['result']['stats']['rcv'])
        loss.append(result['result']['stats']['loss'])
        drop.append(result['result']['stats']['drop'])
        
    for code in countries:
        try:
            country = pycountry.countries.get(alpha_2=code.upper())
            countries_full.append(country.name)
        except AttributeError:
            # If country is not found, append the original code
            countries_full.append(code)

    # Creating a pandas dataframe.
    df = pd.DataFrame({
        'City': cities,
        'Country': countries_full,
        'Region': regions,
        'Continent': continents,
        'Avg': avg_values,
        'Min': min_values,
        'Max': max_values,
        'Tot': total,
        'Rcv': rcv,
        'Los': loss,
        'Dpd': drop
    })

    return df
        
### CLASSES ###
        
class FormInputState(rx.State):
    ping_results: pd.DataFrame = pd.DataFrame()
    loading: bool = False
    show_data: bool = False
    show_initial: bool = True

    @rx.background
    async def handle_submit(self):
        async with self:
            self.loading = True
            self.show_initial = False

        # Fetch ping results in a non-blocking way
        results = await asyncio.to_thread(globalping)

        async with self:
            self.ping_results = results
            self.show_data = True
            self.loading = False
        
    
### MAIN PAGE FUNCTION ###

def globalping_page():
    
    myip_info = myip.info()

    ### CREATING PAGE ###
    
    return rx.fragment(
        # MAIN CONTAINER
        #rx.desktop_only(
        rx.container(
            rx.heading("Get your ping from anywhere! 🌎", size="9", margin_top="10vh", align="left"),
            background_color="#f4f5f8",
            padding="50px 1em 100px 1em"
        ),
        # MINI NAVBAR
        rx.container(
            rx.hstack(
                rx.link("IP Address", 
                        href="/",
                        color="rgba(34,46,58,.5)",
                        background_color="hsla(0,0%,100%,.5)",
                        padding="15px", 
                        border_radius='7px 7px 0px 0px'),
                rx.link("Global Ping", 
                        href="/globalping",
                        color="black",
                        background_color="white", 
                        padding="15px", 
                        border_radius='7px 7px 0px 0px'),
                rx.link("Internet Stats", 
                        href="/internetstats",
                        color="rgba(34,46,58,.5)",
                        background_color="hsla(0,0%,100%,.5)", 
                        padding="15px", 
                        border_radius='7px 7px 0px 0px'),
                background_color="#f4f5f8",
            ),
            background_color="#f4f5f8",
            padding_left="1em",
            padding_right="1em"
        ),
        rx.center(
            rx.container(
                rx.container(height="20px"),
                rx.heading("How does this work?"),
                rx.container(height="20px"),
                rx.text("""
                    Using the awesome features of jsDelivr Globalping, this cool web tool will randomly pick 100 spots across 
                    the globe and gather the ping times (in milliseconds) from your current IP address. To get started, just click "Run Globalping Test." 
                    It only takes around 10 seconds to gather and show your results. Feel free to run it again whenever you want!
                    """,
                    wrap="wrap", width="100%"),
                rx.container(height="10px"),
                rx.text("Every test contains 3 packets and the results are displayed in milliseconds (ms).", wrap="wrap", width="100%"),
                rx.container(height="20px"),
                rx.button(rx.icon(tag="play"), "Run Globalping Test!", on_click=FormInputState.handle_submit, size="3"),
                rx.container(height="20px"),
                rx.cond(
                    FormInputState.loading,
                    rx.chakra.stack(
                    rx.chakra.skeleton_text(
                        no_of_lines=25,
                    ),
                    width="100%",
                ),
                rx.container(height="0px"),
                ),
                rx.cond(
                    FormInputState.show_data,
                    rx.data_table(
                        data=FormInputState.ping_results,
                        search=True,
                        sort=True,
                        resizable=True,
                        pagination=True
                    ),
                rx.cond(
                    FormInputState.show_initial,
                    rx.container(
                        rx.text("Your ping results will be displayed here!", padding="1em", text_align="center"),
                        background_color="#f4f5f8",
                        border_radius="7px",
                        margin_top="20px",
                        width="100%"
                    ),
                    rx.container()
                ),
            ),
            rx.container(height="40px"),
                rx.container(
                    rx.heading("What is ping?"),
                    rx.container(height="20px"),
                    rx.text("""Ping is a measure of the round-trip time it takes for a data packet to travel from the source to the destination and back again. The "good" or acceptable ping range can vary depending on the specific use case and requirements. Here are general categories for ping ranges:""", wrap="wrap", width="100%"),
                    rx.container(height="10px"),
                    rx.unordered_list(
                        rx.container(height="7px"),
                        rx.list_item(
                            rx.icon(
                                "laugh",
                                color="purple",
                                style={"display": "inline"},
                            ),
                            " Low Ping (0-30 ms)",
                        ),
                        rx.container(height="7px"),
                        rx.list_item(
                            rx.icon(
                                "smile",
                                color="green",
                                style={"display": "inline"},
                            ),
                            " Medium Ping (30-100 ms)",
                        ),
                        rx.container(height="7px"),
                        rx.list_item(
                            rx.icon(
                                "meh",
                                color="orange",
                                style={"display": "inline"},
                            ),
                            " High Ping (100-200 ms)",
                        ),
                        rx.container(height="7px"),
                        rx.list_item(
                            rx.icon(
                                "frown",
                                color="red",
                                style={"display": "inline"},
                            ),
                            " Very High Ping (200+ ms)",
                        ),
                        list_style_type="none",
                    ),
                    bg="#f4f5f8",
                    border_radius="7px",
                    padding="1em",
                    wrap="wrap"
                ),
            margin_top="3em",
            margin_bottom="3em",
            width="100%"
            ),
            align="center",
            margin="1em",
        #),
    ),
)